### DISCUSSION TITLE
Render a PDF from a record, a selection of records, or entire collection via a template

### DISCUSSION THREAD

**Discussion by sxsws** [directus/directus#13366] [Feature Requests]
Such uses I'm thinking of include:
- render purchase order data to an invoice
- render a sales order data as a receipt
- render a collection of product records to a catalogue for giving to my customers
- render an entire collection of my_coin_collection records as a pretty hardcopy of data

I'm thinking of the user creating `PDF Templates` consisting of `fields` and then are rendered using a simple template conversion. 

"Simple" templates may include a table for the data which could be created by selecting and dragging the order of columns much like the index view. This table could fit inside another template which the user then somehow creates for the overall layout of the page.

Rather than creating a `PDF Template` creation interface, other solutions may be possible. 
- A pdf design could be created from html (though there are sometimes minor layout differences) which would be easier to create, 
- or the design could be created in something like openoffice and then exported as a pdf, which can then be used as a template.

Image assets could be hard-coded or rendered using image conversion presets or filter strings.

Sub templates may be necessary, e.g. if I wanted to print my entire coin collection, rather than have a pretty table with a small image preview, name of coin, and what year it's from, I may want an entire page devoted to each coin with large images of front, back and side, a detailed explanation about it etc, so the outer template will handle the overall view, whilst a record template would handle each record. etc. This could include a table of contents (which some PDF creation applications like (from memory) html2pdf can create automatically).

Semi  related to https://github.com/directus/directus/discussions/5642 I guess.

**Comment by rijkvanzanten** [comment]
Big fan! Do you have any recommendations / ideas on how to efficiently render the PDF server side? I know there's some approaches using phantomjs or puppeteer to render out PDFs using a virtual/headless browser, but curious to hear your thoughts on other approaches 🙂

**Reply by sxsws** [reply]
wkhtmltopdf is the bog standard opensource solution - https://wkhtmltopdf.org/ - which is available in distributions repos, and can literally take the html and render it to a pdf. It can automatically add a table of contents and has loads of command line options.

It does have its flaws, but it's a start. So it would suffice to render the data to a html file and then pass it to this.

**Reply by rijkvanzanten** [reply]
I'm wondering if that has precompiled builds for all the platforms we're usually available on 🤔

**Reply by sxsws** [reply]
https://wkhtmltopdf.org/downloads.html - just noticed it doesn't look like it's been updated for a while.

**Reply by rijkvanzanten** [reply]
Hmm that's unfortunate. We'd also have to figure out a way how to ship it through `npm` (though there must already be an existing package that ships with it). Something like https://www.npmjs.com/package/pdfkit seems interesting (although that's to render text, not really HTML from what I can tell)

**Reply by sxsws** [reply]
The repo is more recently updated though. I think 'more modern' solutions use something like headless chrome to do it as it handles the rendering more successfully for detailed layouts, but which I think seems rather horrid when something like wkhtmltopdf does it much neater. For data, I think the templates will be rather simple and not need complex css positioning.

Any good? https://www.npmjs.com/search?q=wkhtmltopdf

pdfkit does have an insane amount of downloads, so I'd imagine it's being used in some large project somewhere. No table support though, which is fairly crucial.

https://www.npmjs.com/package/wkhtmltopdf seems to be at least a fairly complete api for it.

**Reply by rijkvanzanten** [reply]
It doesn't seem to render HTML itself though. We will have to use some sort of a browser engine through something like https://www.npmjs.com/package/puppeteer 🤔 

> For data, I think the templates will be rather simple and not need complex css positioning.

I agree! We could come up with an alternative way to configure how it should render, and then effectively have our own PDF rendering engine, instead of having to rely on a full browser rendering engine 🤔

**Reply by sxsws** [reply]
I just wrote a whole edited response but for some reason it didn't save :(

In short: https://github.com/wkhtmltopdf/wkhtmltopdf repo seems recently updated. and https://www.npmjs.com/package/wkhtmltopdf seems like at least a fairly complete api for it.

pdfkit doesn't render tables, which would be a big problem considering lots of data would be tabular.

What do you mean it doesn't render html? Do you mean for the preview from the templates? You could use wkhtml to print a page and then show the preview as an image. Puppeteer seems like a heavy requirement even though it would be an awesome addition.

I'm thinking there could be record-level templates (for things like invoices/quotes) and collection-level templates, with custom options for each conversion. Maybe record-level templates could be built using a graphical drag-and-drop ui html interface like https://github.com/iwangbowen/UI-Builder  or custom templates placed in a directory for specific requirements

**Reply by rijkvanzanten** [reply]
> pdfkit doesn't render tables, which would be a big problem considering lots of data would be tabular.

Ah yeah for real. We'd have to render the lines of the table ourselves then, which feels like a nightmare waiting to happen.

> What do you mean it doesn't render html?

I was thinking you meant using something like `.liquid` for the templates, so you could basically code a HTML page (similar to email templates) that would then be used to generate the PDF. I think that would be the most flexible way of configuring these PDFs, but also require the heaviest dependencies for us to pull of.

**Reply by sxsws** [reply]
Ah, I was thinking of moustache-like templates like `<h1>{{ record.title }}</h1><p>{{ record.description }}</p><ul<li>{{ record.image1 }}</li></ul>` which could then be rendered to html and saved to temp filesl, and wkhtmltopdf could then take that output and convert it to a pdf. The html file has to be taken from a file or even a number of files (from memory, I don't think it can accept piped input) but is doable. If by moustache-like templates, I mean liquid, then yeah, that! I'm new to node, but concepts are similar across languages.

I think you're right, manually drawing tables is a nightmare and shouldn't be entertained! 

I implemented this workflow to get some templated pdf output across a set of records passed into it in symfony projects, I wrote a simple symfony module some time ago. 

I'm starting to think that the user could create something like a 'books' collection that suits their requirements (no hard limits should be imposed for their needs) and consists of fields like title, description, author, maybe keywords, meta description for the pdf, etc. and there could be a special field that allows adding either a single record of certain collection types (like a m2a allows you to specify the allowed types), or a filtered list of a collection type. This whole list and each filtered list of records could be sortable to allow sorting pages. Then any collection with this special field could allow exporting to pdf.

So like this where `MyBookCollection` is a generic collection created by the admin: 

- MyBookCollection:
  - title
  - description
  - author
  - date_created
  - templateText (this contains the liquid template for myBookCollection type - or maybe a m2o to a templates record somehow?) 
  - pagesField: (which has a property of templateText defining that will be applied for each result, or maybe each page could have a templateText property, or a m2o? Also this has a property to define allowed collection types for pages and perhaps other properties like 'add toc', 'pdf metadata' etc) (this list is sortable)
    - singleRecord: SomePageCollection id:1 "Introduction"
    - singleRecord: SomePageCollection: id:5 "About this document"
    - filteredPagesField: SomePageCollection (where tags contains 'chocolate'):  (this could also allow for specifying the templateText for each page result)
      - (results of filtered search that are sortable)
      - singleRecord: SomePageCollection id:11 "Chocolate cup cakes"
      - singleRecord: SomePageCollection id:12 "Chocolate digestives"
      - singleRecord: SomePageCollection id:24 "Chocolate drizzle"
      - singleRecord: SomePageCollection id:33 "Mint Chocolate Biscuits"
      - etc

Converting to pdf would simply be iterating through the pagesField applying templates, building the html file which is then passed to wkhtmltopdf to render the output pdf with appropriate options.

**Reply by rijkvanzanten** [reply]
Sweet! And yeah, the templating engine for the HTML could be whatever really. We're using micromustache for some inline templating things (in-app display templates etc), and rely on `.liquid` for the email templates. Same idea, slightly different syntax, it's all the same really.

I think the only thing to figure out here is how to properly render the rendered `.html` temp file into a `.pdf` in a way that's both flexible and lightweight. Using a full headless browser like Puppeteer feels overkill, but will result in the best possible flexibility (as you could literally do any html/css/js magic to render out the page that's being saved. wkhtmltopdf could be a good alternative, I'm just a little hesitant around any installation quirks that it might bring. One node wrapper I found requires you to install the CLI manually first, which is not gonna fly for a distribution of Directus itself 🤔

**Reply by sxsws** [reply]
I'm a bit of a stickler for simplicity. For things like invoices, catalogues, receipts or whatever uses such data export from directus is going to be used for, I don't really think people will want to do fancy javascript positioning or anything? Maybe you can think of something? From memory I think you can either define inline CSS or use a separate CSS file in the templates for passing to wkhtmltopdf so surely that should be enough? You can do a lot with just CSS, but I'm not sure if it's limited to CSS2 with wkhtmltopdf.

I think there's a system dependency which may or may not need installing depending on the system it's installed on, without which it throws an error when running or doesn't work properly somehow, but doesn't do so during installation. It was a while back and I think I had to search stackoverflow regarding the issue, but it was a simple solution. This was on an older server version of ubuntu so I don't know if it still exists in ubuntu 21. I'm guessing the lack of installation may either be licence-related or perhaps this.

There are other code-only solutions I'm sure, but I think they're going to be hacky, slow and reinvent the wheel. So the two solutions would be headless chromium or wkhtmltopdf I guess. Maybe the latter could be implemented as a start and if necessary, move towards the former for additional features at a later stage. It's going to bring the need for far more server resources though, so it's a big trade to think about. I'm thinking only a small part of the code that actually generates the pdf would need to be modified when the UI to create the 'book' and the post-pdf-file-creation handling code would remain untouched.

If wkhtmltopdf absolutely has to be installed by hand, then the option for pdf export in Directus could be disabled until it's detected that it's installed, and then allow a test where it spits out a sample pdf to check that the installation is fine.

Thinking about it, perhaps there could be various 'PDF Renderers' that are installable by the admin. A code-only option as the default / start renderer, maybe something like pdfkit that is simple to install and get up and running and although doesn't render tables may at some point in the future which could then be pulled upstream. I haven't used pdfkit so it may do some form of rendering data - it does have like half a million weekly downloads! Maybe there's a better pdf node library? And the admin could have a choice of additional available renderers like wkhtmltopdf or puppeteer, but have additional server needs or requirements (which you could *cough* charge more for on your hosted solution?)

Thinking about it even more, who says the export should be limited to just pdf? The 'PDF Renderer' could just be some sort of 'Data Export Renderer' or just 'Renderer' that takes the database data and allows it to be converted to some other format via templates. Any format! PDF / HTML / Text / Markdown /  Images (think infographics) / Whatever someone can think of. Future extensions that people make could hook into this renderer hook to take the data and spit it out in whatever format the user wants. Now that, would be amazing and would require no change to the concept of a pdf export!

If one of the above pageTypes could be say some sort of raw sql-like query, users could compose their own 'page' of data, aggregations, specifically ordered data, etc. It wouldn't matter so long as the template that renders that data knows how to handle the data!

**Comment by Dominic-Marcelino** [comment]
We‘ve implemented a custom solution based on puppertet. It’s working pretty well and the flexibility in general is awesome! 

Performance-wise it’s hard to say for me as we use it to generate PDFs from an external page and only measure the whole process time. From fetching the external html, rendering it in puppertet to create and save the PDF it takes 2-5 seconds for us.

In comparison to most others html to pdf solutions I want to highlight that you get full html/css support this way! Especially tables and flex attributes brought us a lot of trouble with some other solutions.

But just a notice if it’s ever planned to implement in core:
Puppeteer requires some packages that are not installed by default on most systems. Therewith it won’t work in most shared hosting with plesk or cPanel.

**Comment by rijkvanzanten** [comment]
The main concern to me for both solutions presented here is difficulty of installation. So far, Directus doesn't really require anything to be installed system wide to run, as it's all NodeJS[^1].

Once we start including more specific non-node packages that are required for running Directus, we might have to enforce usage of Docker as the primary installation method (rather than NodeJS), as that would be the only we can reliable guarantee the availability of all required libraries 

[^1]: Heavy asterisk: technically argon2 requires openssl, but it seems like it's generally reliably available in platforms we're installing in.

**Reply by alexbrowng** [reply]
I think that being able to generate a PDF of an element of a collection through a template is a great functionality and I agree with @rijkvanzanten that it can't make the directus installation process worse, but, I propose a different approach, in the end it is to use all the pieces you have proposed, implement a system that allows to create HTML/CSS/JS templates and render them with the variables of an element of a collection or a set of elements of a collection based on a filter to display them from the administration panel as a new window, thus allowing the export to PDF from the user's browser, so you avoid installing a tool to convert HTML to PDF and also, you allow the forms of printing that the browser supports and could be a new way to "preview" in a more "beautiful" and customizable format than a form.

**Reply by rijkvanzanten** [reply]
Ahh that's smart!

**Reply by alexbrowng** [reply]
It could also be useful in addition to the sharing feature

**Reply by sxsws** [reply]
why not have a default 'Renderer' (from my comment above) that renders to html and leave the pdf export to an extension that's available through the marketplace?

**Comment by loongmxbt** [comment]
Mark, very useful feature, seems to need a custom extension.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by halwesit** [reply]
it has 23 votes!

**Reply by rijkvanzanten** [reply]
It didn't on May 10, 2023

**Reply by halwesit** [reply]
Yeah, you're right, I apologize, may work on it now?

**Comment by atipezda** [comment]
+1 i will really apreciate pdf generation in directus

**Reply by sharkfin009** [reply]
I have a need to render cards (which contain lots of text manipulation and flexbox) from my front end so I can build a grid of PNGS for an email template. Trying to figure out how to get Puppeteer working in docker.  I like @alexbrowng 's idea... in my situation it would help a lot

**Reply by sharkfin009** [reply]
ok this made me think, i could build a module extension that uses my card.vue file from my front end to render a grid of cards . Then use something like  https://github.com/tsayen/dom-to-image to render pngs for use in the email template.  Lol sorry this is me talking it through. So this approach is  basically building an email ( or PDF ) preview tool that , when the user is happy with the outcome ( newsletter or whatever ) , and clicks send , then a liquid template gets built including elements that need to be rendered to images, like cards or anything that wont work (flexbox! ) in the template
